### PR TITLE
chgrp: replace getopts with clap (#2118)

### PR DESCRIPTION
**Pull request by wfscheper**
Hey, I'm trying to learn rust and this looked like something straightforward to do, especially since the `chown` command is so similar.

I'm not 100% certain that the I have the best way to handle the `--reference` flag, but I couldn't come up with any options that left the `files` Vec immutable.

### CODE REVIEW

**Comment by jhscheer** [comment]
> I'm not 100% certain that the I have the best way to handle the `--reference` flag, but I couldn't come up with any options that left the `files` Vec immutable.

An alternative could be to not use `ARG_GROUP` and instead only use `ARG_FILES` and then skip the first file if `--reference` is not present.
What do you think (I also changed some minor nitpicks)?
```diff
diff --git a/src/uu/chgrp/src/chgrp.rs b/src/uu/chgrp/src/chgrp.rs
index 7578ec9f..2c19446a 100644
--- a/src/uu/chgrp/src/chgrp.rs
+++ b/src/uu/chgrp/src/chgrp.rs
@@ -51,18 +51,16 @@ pub mod options {
         pub static EVERY: &str = "L";
     }
     pub static REFERENCE: &str = "reference";
+    pub static ARG_FILES: &str = "files";
 }
 
-static ARG_GROUP: &str = "group";
-static ARG_FILES: &str = "files";
-
 const FTS_COMFOLLOW: u8 = 1;
 const FTS_PHYSICAL: u8 = 1 << 1;
 const FTS_LOGICAL: u8 = 1 << 2;
 
 fn get_usage() -> String {
     format!(
-        "{0} [OPTION]... GROUP FILE...\n{0} [OPTION]... --reference=RFILE FILE...",
+        "{0} [OPTION]... GROUP FILE...\n    {0} [OPTION]... --reference=RFILE FILE...",
         executable!()
     )
 }
@@ -126,8 +124,9 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
             Arg::with_name(options::REFERENCE)
                 .long(options::REFERENCE)
                 .value_name("RFILE")
-                .help("use RFILE's owner and group rather than specifying OWNER:GROUP values")
-                .min_values(1),
+                .help("use RFILE's group rather than specifying a GROUP value")
+                .takes_value(true)
+                .multiple(false),
         )
         .arg(
             Arg::with_name(options::RECURSIVE)
@@ -152,28 +151,34 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 .help("traverse every symbolic link to a directory encountered"),
         )
         .arg(
-            Arg::with_name(ARG_GROUP).value_name(ARG_GROUP)
-                .multiple(false)
-                .takes_value(true)
-                .required(true),
-        )
-        .arg(
-            Arg::with_name(ARG_FILES)
-                .value_name(ARG_FILES)
+            Arg::with_name(options::ARG_FILES)
+                .value_name(options::ARG_FILES)
                 .multiple(true)
                 .takes_value(true)
-                .required(false)
+                .required(true)
                 .min_values(1),
         )
         .get_matches_from(args);
 
     /* First arg is the group */
-    let group = matches.value_of(ARG_GROUP).unwrap();
+    let group: &str = matches
+        .values_of(options::ARG_FILES)
+        .unwrap()
+        .next()
+        .unwrap();
 
     /* Then the list of files */
-    let mut files: Vec<String> = matches
-        .values_of(ARG_FILES)
-        .map(|v| v.map(ToString::to_string).collect())
+    let files: Vec<String> = matches
+        .values_of(options::ARG_FILES)
+        .map(|v| {
+            v.map(ToString::to_string)
+                .skip(if matches.is_present(options::REFERENCE) {
+                    0
+                } else {
+                    1
+                })
+                .collect()
+        })
         .unwrap_or_default();
 
     let preserve_root = matches.is_present(options::preserve_root::PRESERVE);
@@ -221,8 +226,6 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
     if let Some(file) = matches.value_of(options::REFERENCE) {
         match fs::metadata(&file) {
             Ok(meta) => {
-                // "group" is really a file
-                files.push(group.to_string());
                 dest_gid = meta.gid();
             }
             Err(e) => {

```

**Comment by wfscheper** [comment]
I did have the thought that I could parse `args` twice, or use some other method to check if the `--reference` option was used. If it was used then parse with just `ARG_FILES`, otherwise with both `ARG_GROUP` and `ARG_FILES`. If you don't setup the clap parser correctly you end up with error messages that don't line up with the usage like this:
```bash
$ chgrp --reference=/etc/passwd
error: The following required arguments were not provided:
    <group>

USAGE:
    chgrp [OPTION]... GROUP FILE...
    chgrp [OPTION]... --reference=RFILE FILE...

For more information try --help
```

**Comment by jhscheer** [comment]
The approach above prints:
```bash
$ target/debug/chgrp --reference=/etc/passwd
error: The following required arguments were not provided:
    <files>...

USAGE:
    chgrp [OPTION]... GROUP FILE...
    chgrp [OPTION]... --reference=RFILE FILE...
```
which is just fine.
However, if called with only one argument, nothing is printed, although only GROUP is provided and FILE is missing.
This is tricky and I couldn't find a way to solve this using just clap's (v2.33) various `Arg::required*` methods.

File: src/uu/chgrp/src/chgrp.rs [hunk 1]
```diff
@@ -24,76 +26,173 @@ use std::os::unix::fs::MetadataExt;
...
+    pub mod traverse {
+        pub static TRAVERSE: &str = "H";
+        pub static NO_TRAVERSE: &str = "P";
+        pub static EVERY: &str = "L";
+    }
+    pub static REFERENCE: &str = "reference";
+}
+
+static ARG_GROUP: &str = "group";
+static ARG_FILES: &str = "files";
```
Comment (jhscheer) [inline src/uu/chgrp/src/chgrp.rs:57]: Please move to `mod options`

File: src/uu/chgrp/src/chgrp.rs [hunk 2]
```diff
@@ -24,76 +26,173 @@ use std::os::unix::fs::MetadataExt;
...
 
 const FTS_COMFOLLOW: u8 = 1;
 const FTS_PHYSICAL: u8 = 1 << 1;
 const FTS_LOGICAL: u8 = 1 << 2;
 
+fn get_usage() -> String {
+    format!(
+        "{0} [OPTION]... GROUP FILE...\n{0} [OPTION]... --reference=RFILE FILE...",
```
Comment (jhscheer) [inline src/uu/chgrp/src/chgrp.rs:65]: second line is not alined with first line

File: src/uu/chgrp/src/chgrp.rs [hunk 3]
```diff
@@ -24,76 +26,173 @@ use std::os::unix::fs::MetadataExt;
...
+        .arg(
+            Arg::with_name(options::preserve_root::NO_PRESERVE)
+                .long(options::preserve_root::NO_PRESERVE)
+                .help("do not treat '/' specially (the default)"),
+        )
+        .arg(
+            Arg::with_name(options::REFERENCE)
+                .long(options::REFERENCE)
+                .value_name("RFILE")
+                .help("use RFILE's owner and group rather than specifying OWNER:GROUP values")
```
Comment (jhscheer) [inline src/uu/chgrp/src/chgrp.rs:129]: GNU's chgrp help says: "use RFILE's group rather than specifying a GROUP value"

File: src/uu/chgrp/src/chgrp.rs (hunk 1 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs]: Done

File: src/uu/chgrp/src/chgrp.rs (hunk 2 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs]: Done

File: src/uu/chgrp/src/chgrp.rs (hunk 3 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs]: Done

**Comment by wfscheper** [comment]
I fixed the lines you cited, and pushed a commit that is the only way I can figure out how to change the number of positional arguments depending on the `--reference` option. I'm not very pleased with it, but there doesn't appear to be any way to do this with a single clap App instance.

**Comment by wfscheper** [comment]
Probably the best solution is to do what you suggested, and handle the 1 positional argument situation in chgrp instead of relying on clap.

**Comment by wfscheper** [comment]
I figured out how to do this after poking around in clap's issue tracker: https://docs.rs/clap/2.33.3/clap/enum.AppSettings.html#variant.AllowMissingPositional

**Comment by jhscheer** [comment]
> I figured out how to do this after poking around in clap's issue tracker: https://docs.rs/clap/2.33.3/clap/enum.AppSettings.html#variant.AllowMissingPositional

Nice, I didn't know about this. That could also be useful for other uutils.

There are still some not optimal corner cases.
Option `Arg::require_equals` for `options::REFERENCE` helps, but doesn't solve all of them.
Do you have an idea?

```bash
$ target/debug/coreutils chgrp --reference ref_file
error: The following required arguments were not provided:
    <files>...

$ target/debug/coreutils chgrp -v --reference ref_file file1
chgrp: group of 'file1' retained as staff

$ target/debug/coreutils chgrp -v --reference=ref_file file1 file2
chgrp: group of 'file1' retained as staff
chgrp: group of 'file2' retained as staff

$ target/debug/coreutils chgrp -v --reference ref_file file1 file2
error: The following required arguments were not provided:
    <files>...
```
The first 3 look good but the last is not right.

```bash
target/debug/coreutils chgrp -v file1 file2 file3 --reference ref_file
chgrp: group of 'file2' retained as staff
chgrp: group of 'file3' retained as staff
```
Here `file1` is missing.

```bash
$ target/debug/coreutils chgrp groupname
error: The following required arguments were not provided:
    <group>
```
Here FILE is missing, but clap says GROUP is missing.

**Comment by tertsdiepraam** [comment]
```
$ target/debug/coreutils chgrp -v --reference ref_file file1 file2
error: The following required arguments were not provided:
    <files>...
```
This one seems to be fixed when I change `min_values(1)` to `takes_value(true)` for `--reference`. The other ones are harder but maybe we could check whether `--reference` is in `args` and only add the `group` argument to `clap` if it isn't? It's ugly, but I also don't see a way to do this only through `clap`. At least you get the correct error messages for missing arguments this way.

So something like this:
```diff
diff --git a/src/uu/chgrp/src/chgrp.rs b/src/uu/chgrp/src/chgrp.rs
index 3c03ff96..d0048468 100644
--- a/src/uu/chgrp/src/chgrp.rs
+++ b/src/uu/chgrp/src/chgrp.rs
@@ -14,7 +14,6 @@ use uucore::fs::resolve_relative_path;
 use uucore::libc::gid_t;
 use uucore::perms::{wrap_chgrp, Verbosity};
 
-use clap::AppSettings::AllowMissingPositional;
 use clap::{App, Arg};
 
 extern crate walkdir;
@@ -74,8 +73,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
 
     let usage = get_usage();
 
-    let matches = App::new(executable!())
-        .setting(AllowMissingPositional)
+    let mut app = App::new(executable!())
         .version(VERSION)
         .about(ABOUT)
         .usage(&usage[..])
@@ -128,7 +126,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 .long(options::REFERENCE)
                 .value_name("RFILE")
                 .help("use RFILE's group rather than specifying GROUP values")
-                .min_values(1),
+                .takes_value(true),
         )
         .arg(
             Arg::with_name(options::RECURSIVE)
@@ -151,23 +149,30 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
             Arg::with_name(options::traverse::EVERY)
                 .short(options::traverse::EVERY)
                 .help("traverse every symbolic link to a directory encountered"),
-        )
-        .arg(
+        );
+
+    if !args
+        .iter()
+        .any(|s| s == "--reference" || s.starts_with("--reference="))
+    {
+        app = app.arg(
             Arg::with_name(options::ARG_GROUP)
                 .value_name(options::ARG_GROUP)
                 .multiple(false)
                 .takes_value(true)
-                .required_unless(options::REFERENCE)
-        )
-        .arg(
-            Arg::with_name(options::ARG_FILES)
-                .value_name(options::ARG_FILES)
-                .multiple(true)
-                .takes_value(true)
-                .required(true)
-                .min_values(1),
-        )
-        .get_matches_from(args);
+                .required(true),
+        );
+    }
+    app = app.arg(
+        Arg::with_name(options::ARG_FILES)
+            .value_name(options::ARG_FILES)
+            .multiple(true)
+            .takes_value(true)
+            .required(true)
+            .min_values(1),
+    );
+
+    let matches = app.get_matches_from(args);
 
     /* Get the list of files */
     let files: Vec<String> = matches
```

**Comment by wfscheper** [comment]
The `args.iter().any(...)` solution will fail on the pathological case of a file named `--reference` or `--reference=...`, which should work with correct usage of `--` to stop parsing of options (I verified that gnu chgrp and uutils chgrp from master behave this way). 

I'm still trying to address @jhscheer 's edge cases with the AllowMissingPositional setting. Clearly `.required_unless()` doesn't do anything with that setting turned on.

**Comment by tertsdiepraam** [comment]
Ah right, good catch! To be honest, I'm not sure it's possible with `AllowMissingPositional`, it seems to be there for slightly different use cases, but it'd be awesome if you do figure it out. I think my solution can be fixed with `.iter().take_while(|s| s != "--").any(...)` or by using a for loop (also fixing the `--help` issue):
```rust
let group = true;
for s in args {
    if s == "--reference" || s.starts_with("--reference=") {
        group = false;
    } else if s == "-h || s == "--help" {
        group = true;
        break;
    } else if s == "--" {
        break;
    }
}
```
I'm on mobile so it's probably not quite right :)

**Comment by sylvestre** [comment]
@wfscheper do you have enough information to finish this PR? thanks

**Comment by wfscheper** [comment]
Yes, I think so. Work caught up with me, but hopefully I’ll get back to it
this weekend.

**Comment by wfscheper** [comment]
Okay, using `.iter().take_while()` passes all of the edge case tests around `--reference`. I do have to check args twice if you want `--help` to always show both args, unless some one knows of a way to do that in one iteration?

File: src/uu/chgrp/src/chgrp.rs [hunk 4]
```diff
@@ -24,76 +26,197 @@ use std::os::unix::fs::MetadataExt;
...
+                .help("do not traverse any symbolic links (default)")
+                .overrides_with_all(&[options::traverse::TRAVERSE, options::traverse::EVERY]),
+        )
+        .arg(
+            Arg::with_name(options::traverse::EVERY)
+                .short(options::traverse::EVERY)
+                .help("traverse every symbolic link to a directory encountered"),
+        );
+
+    let mut reference = false;
```
Comment (tertsdiepraam) [inline src/uu/chgrp/src/chgrp.rs:155]: Could you add a comment explaining this?

File: src/uu/chgrp/src/chgrp.rs [hunk 5]
```diff
@@ -106,27 +229,20 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
...
 
-    if matches.free.is_empty() {
-        show_usage_error!("missing operand");
-        return 1;
-    } else if matches.free.len() < 2 && !matches.opt_present("reference") {
-        show_usage_error!("missing operand after ‘{}’", matches.free[0]);
-        return 1;
-    }
-
-    let dest_gid: gid_t;
```
Comment (tertsdiepraam) [inline src/uu/chgrp/src/chgrp.rs:127]: What was you reason to change this to `u32`?

File: src/uu/chgrp/src/chgrp.rs [hunk 6]
```diff
@@ -24,76 +26,197 @@ use std::os::unix::fs::MetadataExt;
...
+            Arg::with_name(options::traverse::EVERY)
+                .short(options::traverse::EVERY)
+                .help("traverse every symbolic link to a directory encountered"),
+        );
+
+    let mut reference = false;
+    for arg in args.iter().take_while(|s| *s != "--") {
+        if arg.starts_with("--reference=") || arg == "--reference" {
+            reference = true;
+            break;
```
Comment (tertsdiepraam) [inline src/uu/chgrp/src/chgrp.rs:159]: If you break here, I think that the combination `--reference --help` is not quite right. Although `--help --reference` does work. (Feel free to correct me if it does work)

File: src/uu/chgrp/src/chgrp.rs [hunk 7]
```diff
@@ -24,76 +26,197 @@ use std::os::unix::fs::MetadataExt;
...
+        } else if arg == "--help" {
+            break;
         }
     }
 
-    let matches = opts.parse(args);
-    let recursive = matches.opt_present("recursive");
+    let matches: ArgMatches;
+    if reference {
+        matches = app
```
Comment (tertsdiepraam) [inline src/uu/chgrp/src/chgrp.rs:167]: Might be better to make `app` mutable and do `app = app.arg(...)` so that you don't to repeat the files argument.

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Awesome! I think `chgrp --reference --help` still hides the group argument. It would be great if you could fix that.

File: src/uu/chgrp/src/chgrp.rs (hunk 4 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs:157]: Done

File: src/uu/chgrp/src/chgrp.rs (hunk 5 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs:127]: Hmm. I probably cribbed that from chown. I can revert it if need be.

File: src/uu/chgrp/src/chgrp.rs (hunk 6 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs]: Good catch, I thought I tested that but clearly not.

File: src/uu/chgrp/src/chgrp.rs (hunk 7 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs]: Done.

**Review by tertsdiepraam — APPROVED** [review]
Thanks! It looks good now

File: src/uu/chgrp/src/chgrp.rs (hunk 5 above)
Comment (tertsdiepraam) [reply src/uu/chgrp/src/chgrp.rs:127]: Looking at `libc`, it seems that `gid_t` is always `u32`. The `gid` function returns `u32`, so I think changing it to u32 makes sense.

**Comment by tertsdiepraam** [comment]
Uh oh, Ubuntu seems to fail:
```
---- test_chgrp::test_big_p stdout ----
current_directory_resolved: 
run: /target/i686-unknown-linux-gnu/debug/coreutils chgrp -RP bin /proc/self/cwd
thread 'test_chgrp::test_big_p' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"chgrp: changing group of \'/proc/self/cwd\': Operation not permitted (os error 1)\nchgrp: changing group of \'/proc/self/cwd/file3\': Operation not permitted (os error 1)\nchgrp: changing group of \'/proc/self/cwd/file1\': Operation not permitted (os error 1)\nchgrp: changing group of \'/proc/self/cwd/ref_file\': Operation not permitted (os error 1)\nchgrp: changing group of \'/proc/self/cwd/file2\': Operation not permitted (os error 1)"
>"chgrp: changing group of \'/proc/self/cwd\': Operation not permitted (os error 1)"

', tests/common/util.rs:243:9

---- test_chgrp::test_reference_last stdout ----
current_directory_resolved: 
run: /target/i686-unknown-linux-gnu/debug/coreutils chgrp -v file1 file2 file3 --reference ref_file
thread 'test_chgrp::test_reference_last' panicked at 'Command was expected to succeed.
stdout = 
 stderr = thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: NotFound, error: "No such id: 116" }', /project/src/uucore/src/lib/features/perms.rs:95:44
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
', tests/common/util.rs:139:13

---- test_chgrp::test_reference_multi_no_equal stdout ----
current_directory_resolved: 
run: /target/i686-unknown-linux-gnu/debug/coreutils chgrp -v --reference ref_file file1 file2
thread 'test_chgrp::test_reference_multi_no_equal' panicked at 'Command was expected to succeed.
stdout = 
 stderr = thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: NotFound, error: "No such id: 116" }', /project/src/uucore/src/lib/features/perms.rs:95:44
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
', tests/common/util.rs:139:13
```

File: src/uu/chgrp/src/chgrp.rs [hunk 8]
```diff
@@ -24,76 +26,192 @@ use std::os::unix::fs::MetadataExt;
...
         }
     }
 
-    let matches = opts.parse(args);
-    let recursive = matches.opt_present("recursive");
+    if help || !reference {
+        // add both positional arguments
+        app = app.arg(
+            Arg::with_name(options::ARG_GROUP)
+                .value_name(options::ARG_GROUP)
```
Comment (jhscheer) [inline src/uu/chgrp/src/chgrp.rs:175]: should be "GROUP"

File: src/uu/chgrp/src/chgrp.rs [hunk 9]
```diff
@@ -24,76 +26,192 @@ use std::os::unix::fs::MetadataExt;
...
+                .required(true)
+                .takes_value(true)
+                .multiple(false),
+        )
+    }
+    app = app.arg(
+        Arg::with_name(options::ARG_FILES)
+            .value_name(options::ARG_FILES)
```
Comment (jhscheer) [inline src/uu/chgrp/src/chgrp.rs:183]: should be "FILE"

**Comment by wfscheper** [comment]
Should I change the value of ARG_GROUP and ARG_FILES, or do you want the
value_name set separately?

**Comment by jhscheer** [comment]
> Should I change the value of ARG_GROUP and ARG_FILES, or do you want the value_name set separately?

I don't really mind, imho. the former is more elegant.

**Comment by wfscheper** [comment]
I'm not able to reproduce the ubuntu failures with a docker container, so I'm not sure how best to go about fixing those errors.

File: src/uu/chgrp/src/chgrp.rs (hunk 8 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs:175]: Done

File: src/uu/chgrp/src/chgrp.rs (hunk 9 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs:183]: Done

File: src/uu/chgrp/src/chgrp.rs [hunk 10]
```diff
@@ -24,76 +26,192 @@ use std::os::unix::fs::MetadataExt;
...
+        .map(|v| v.map(ToString::to_string).collect())
+        .unwrap_or_default();
+
+    let preserve_root = matches.is_present(options::preserve_root::PRESERVE);
+
+    let mut derefer = if matches.is_present(options::dereference::NO_DEREFERENCE) {
+        1
+    } else {
+        0
+    };
```
Comment (tertsdiepraam) [inline src/uu/chgrp/src/chgrp.rs:204]: I think this is backwards. Also note that it doesn't match the previous implementation and the GNU `chgrp`. The previous implementation has a default `derefer` of `-1`, because the default depends on whether the `-R` flag is given:
```
--dereference
              affect the referent of each symbolic link (this is the default), rather than the symbolic link itself

       -h, --no-dereference
              affect symbolic links instead of any referenced file (useful only on systems that can change the ownership of a symlink)
...
...
...
The  following options modify how a hierarchy is traversed when the -R option is also specified.  If more than one is specified, only the
       final one takes effect.

       -H     if a command line argument is a symbolic link to a directory, traverse it

       -L     traverse every symbolic link to a directory encountered

       -P     do not traverse any symbolic links (default)
```

As an example:
```
chgrp -R --dereference bin /proc/self/cwd  # this fails
chgrp -R bin /proc/self/cwd                # this succeeds
chgrp -RP bin /proc/self/cwd               # this succeeds and is equivalent to the one above
```
I think solving this would fix at least `test_big_p`

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
I think I found the issue for one of the failing tests. I'll keep looking for the others.

File: src/uu/chgrp/src/chgrp.rs (hunk 10 above)
Comment (wfscheper) [reply src/uu/chgrp/src/chgrp.rs:204]: That's another bit I lifted from the chown command, probably should've thought through the changes more and noticed that the values were significantly different. Fixed in the latest push.

Comment (tertsdiepraam) [reply src/uu/chgrp/src/chgrp.rs]: Awesome! Let's see what the CI does now!

**Comment by tertsdiepraam** [comment]
I have a guess for the other failing tests. It seems to fail on trying to format the message `"group of '{}' retained as {}"` because it cannot find the group (probably). Here is the easiest fix I can think of (in `src/uucore/src/lib/features/perms.rs`):
```rust
out = format!(
    "group of '{}' retained as {}",
    path.display(),
    entries::gid2grp(dest_gid).unwrap() // change this to .unwrap_or_default()
);
```
The output will then be wrong (because the group will be an empty string), but it will succeed and pass the test. In other PR's we should then make `wrap_chgrp` return errors instead of panicking and figure out why the CI cannot find these groups.

**Comment by wfscheper** [comment]
Sounds reasonable, I'll try and update the PR tonight.

**Comment by tertsdiepraam** [comment]
And I think the `test_big_p` is just broken. It fails on master when I add a file in the fixtures, so changing the `stderr_is` to `stderr_contains` should be fine and I'll make another issue for that (because the behaviour does not match GNU)

**Comment by wfscheper** [comment]
One more time...

Changed `test_reference` to use a fixture file as the referent, that way the groups don't match.

**Review by tertsdiepraam — APPROVED** [review]

**Comment by tertsdiepraam** [comment]
Thanks again! Great work on this PR!

**Comment by wfscheper** [comment]
Thanks! Would it make sense to make the same sort of ‘—reference’ changes
to chmod next, or pick up one of the other clap conversion issues?

**Comment by tertsdiepraam** [comment]
Definitely! There are probably some other arguments in other utils that aren't quite right, so fixing some of those would be great! I've also noticed that there might be some unnecessary `OsStr` to `String` conversions. Ideally, we would use `value_of_os` and `values_of_os` as much as possible. You can just pick anything you like :)
